wasmplugin/plugin.go

```diff
@@ -357,20 +358,27 @@ func (ctx *httpContext) OnHttpStreamDone() {
 	logMemStats()
 }
 
-func (ctx *httpContext) handleInterruption(phase string, interruption *ctypes.Interruption) types.Action {
-	ctx.metrics.CountTXInterruption(phase, interruption.RuleID)
+const noGRPCStream int32 = -1
 
-	proxywasm.LogInfof("%d interrupted, action %q", ctx.contextID, interruption.Action)
-	statusCode := interruption.Status
-	if statusCode == 0 {
-		statusCode = 403
-	}
+func (ctx *httpContext) handleInterruption(phase string, interruption *ctypes.Interruption) types.Action {
+	if !ctx.triggeredInterruption {
+		ctx.triggeredInterruption = true
+		ctx.metrics.CountTXInterruption(phase, interruption.RuleID)
+
+		proxywasm.LogInfof("%d interrupted, action %q, during phase %s", ctx.contextID, interruption.Action, phase)
+		statusCode := interruption.Status
+		if statusCode == 0 {
+			statusCode = 403
+		}
+		if err := proxywasm.SendHttpResponse(uint32(statusCode), nil, nil, noGRPCStream); err != nil {
+			panic(err)
+		}
 
-	if err := proxywasm.SendHttpResponse(uint32(statusCode), nil, nil, -1); err != nil {
-		panic(err)
+		return types.ActionPause
+	} else {
+		proxywasm.LogDebug("interruption already triggered")
+		return types.ActionContinue
```

Comment:
Not sure this is the right fix, this means that after an intervention being handled we will only return `ActionContinue`, however, the SendHttpResponse documentation says:

```
After you've invoked this function, you *must* return types.Action.Pause to
// stop further processing of the initial HTTP request/response.
```

But I think that was what lead us to the blocking on the requests cc @mathetake

Reply:
Yes, definitely, it is an open point, and I wanted to elaborate on it before removing the draft status. Thanks for it.

Reply:
yeah if you continue and the upstream replies the response, then  your client _might_ receive response twice and might crash. (Also, I am not sure if Envoy won't crash either 😉)

Reply:
also if you continue, your upstream will receive this malicious (?) body, is that what you want?

Reply:
I came up with the following reasoning:
When we want to Deny a request, we rely on `handleInterruption`, which sends back a response (usually `403`) thanks to `proxywasm.SendHttpResponse(uint32(statusCode), nil, nil, noGRPCStream)` followed by `ActionPause`. 

The `ActionPause` will stop the malicious request, so `onHttpRequestBody` is not going to be called and nothing will be sent upstream.

Internally `SendHttpResponse` calls [`internal.ProxySendLocalResponse`](https://github.com/envoyproxy/envoy/blob/b7fb1da00951e14c8fdcefdfc2c44d64ea055613/source/extensions/common/wasm/context.cc#L1606) that calls [`DownstreamFilterManager::sendLocalReply`](https://github.com/envoyproxy/envoy/blob/829ed5d51fb27ccf83ed3acd0309037f4c9fc1ac/source/common/http/filter_manager.cc#L852).
Here two things may happen depending on when the interruption is raised:
- If it happens during `OnHttpRequestHeaders/Body`: the generated response will be still **send via the filter chain** (calling `sendLocalReplyViaFilterChain`). It leads to trigger the `OnHttpResponseHeaders`, because it actually is a response that is coming back (Here happens the second interruption that we are experiencing wasm filter side). To the best of my knowledge, letting this response go with `ActionContinue` just means: Send downstream (to the client) the 403 response.
- If it happens during `OnHttpResponseHeaders`: the response is **not** sent back via the filter chain (Envoy calls `sendDirectLocalReply` instead of  `sendLocalReplyViaFilterChain`). This is when we are experiencing just one interruption. Indeed the requests when it happens are, for example:
   - `curl -I 'http://localhost:8080/status/406'`: the rule triggers a 403 response looking at the response headers
   - `curl -I 'http://localhost:8080/anything?arg=<script>alert(0)</script>'`: yet another problem, but in this case, there is no request body, so OnRequestBody is not triggered. The XSS payload is just detected (too late) during OnResponseHeaders.
   
We can see it also looking at the Envoy logs:
When we experience the single handleInterruption call, there is just one line and `directly to the encoder` means that we are skipping the filter chain:
```
[source/common/http/filter_manager.cc:882] [C4][S877706587745125579] Sending local reply with details  directly to the encoder
```
When we experience the double handleInterruption call: we have two lines (running Envoy 1.23). The first is logged by `sendLocalReplyViaFilterChain` and the second by `sendDirectLocalReply`:
```
[source/common/http/filter_manager.cc:905] [C6][S1165633363438994845] Sending local reply with details
[source/common/http/filter_manager.cc:882] [C4][S877706587745125579] Sending local reply with details  directly to the encoder
```

If this reasoning is right:
- We may just use `ActionContinue` maybe checking that the response code is the one expected by the interruption?
- Find a way to always send the response directly and not via the filter chain. From them Envoy codebase I see a flag named `avoid_reentrant_filter_invocation_during_local_reply_`, but I don't know who is in charge to set it and if it is doable form the SDK